src/utils/distributed-state-registry.js

```diff
@@ -0,0 +1,411 @@
+	 * @returns {void}
+	 */
+	_remove( name, serverName ) {
+		var name, exists = false;
+
+		if( !this._data[ name ] ) {
+			return;
+		}
+
+		this._data[ name ].nodes[ serverName ] = false;
+
+		for( serverName in this._data[ name ].nodes ) {
+			if( this._data[ name ].nodes[ serverName ] === true ) {
+				exists = true;
+			}
+		}
+
+		if( exists === false ) {
+			delete this._data[ name ];
+			this.emit( 'remove', name );
+		}
+	}
+
+	/**
+	 * Adds a new entry to this registry, either as a result of a remote or
+	 * a local addition. Will emit an `add` event if the entry wasn't present before
+	 *
+	 * @param {String} name       the name of the new entry
+	 * @param {String} serverName the name of the server that added the entry
+	 *
+	 * @private
+	 * @returns {void}
+	 */
+	_add( name, serverName ) {
+		if( !this._data[ name ] ) {
+			this._data[ name ] = {
+				nodes: {},
+				checkSum: this._createCheckSum( name )
+			}
+			this.emit( 'add', name );
+		}
+
+		this._data[ name ].nodes[ serverName ] = true;
+	}
+
+	/**
+	 * Checks if a given entry exists for a given serverName
+	 *
+	 * @param   {String}  name       the name of the entry
+	 * @param   {String}  serverName the name of the server that might hold the entry
+	 *
+	 * @private
+	 * @returns {Boolean} exists
+	 */
+	_has( name, serverName ) {
+		return this._data[ name ] && this._data[ name ].nodes[ serverName ];
+	}
+
+	/**
+	 * Generic messaging function for add and remove messages
+	 *
+	 * @param   {String} name   the name of the entry that's added or removed
+	 * @param   {String} action action-constant, one of C.EVENT.<>
+	 *
+	 * @private
+	 * @returns {void}
+	 */
+	_sendMessage( name, action ) {
+		var message = {
+			topic: this._topic,
+			action: action,
+			data: [ name, this._options.serverName, this._getCheckSumTotal( this._options.serverName ) ]
+		};
+
+		this._options.messageConnector.publish( this._topic, message );
+	}
+
+	/**
+	 * This method calculates the total checkSum for all local entries of
+	 * a given serverName
+	 *
+	 * @param {String} serverName the name of the server for which the checkSum should be calculated
+	 *
+	 * @private
+	 * @returns {Number} totalCheckSum
+	 */
+	_getCheckSumTotal( serverName ) {
+		var totalCheckSum = 0, name;
+
+		for( name in this._data ) {
+			if( this._data[ name ].nodes[ serverName ] ) {
+				totalCheckSum += this._data[ name ].checkSum;
+			}
+		}
+
+		return totalCheckSum;
+	}
+
+	/**
+	 * Calculates a simple checkSum for a given name. This is done up-front and cached
+	 * to increase performance for local add and remove operations. Arguably this is a generic
+	 * method and might be moved to the utils class if we find another usecase for it.
+	 *
+	 * @param   {String} name the name of the entry
+	 *
+	 * @private
+	 * @returns {Number} checkSum
+	 */
+	_createCheckSum( name ) {
+		var checkSum = 0, i;
+
+		for( i = 0; i < name.length; i++ ) {
+			checkSum += name.charCodeAt( i );
+		}
+
+		return checkSum;
+	}
```

Comment:
yup, but we're using the sum of the checksums to get a rough impression of state-consistency. This happens every time something is added to the registry, so its a trade off between performance and uniqueness. Happy to go for something more elaborate though if you've got an algorithm in mind that doesn't come with significant performance overhead